csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedCondition.qll

```diff
@@ -0,0 +1,251 @@
+    this = TTranslatedNativeCondition(expr)
+  }
+
+  override final Instruction getChildSuccessor(TranslatedElement child) {
+    none()
+  }
+}
+
+abstract class TranslatedBinaryLogicalOperation extends
+  TranslatedNativeCondition, ConditionContext {
+  override BinaryLogicalOperation expr;
+
+  override final TranslatedElement getChild(int id) {
+    id = 0 and result = getLeftOperand() or
+    id = 1 and result = getRightOperand()
+  }
+
+  override final Instruction getFirstInstruction() {
+    result = getLeftOperand().getFirstInstruction()
+  }
+
+  override final predicate hasInstruction(Opcode opcode, InstructionTag tag,
+    Type resultType, boolean isLValue) {
+    none()
+  }
+
+  override final Instruction getInstructionSuccessor(InstructionTag tag,
+    EdgeKind kind) {
+    none()
+  }
+
+  final TranslatedCondition getLeftOperand() {
+    result = getTranslatedCondition(expr.getLeftOperand())
+  }
+
+  final TranslatedCondition getRightOperand() {
+    result = getTranslatedCondition(expr.getRightOperand())
+  }
+}
+
+class TranslatedLogicalAndExpr extends TranslatedBinaryLogicalOperation {
+  TranslatedLogicalAndExpr() {
+    expr instanceof LogicalAndExpr
+  }
+
+  override Instruction getChildTrueSuccessor(TranslatedCondition child) {
+    (
+      child = getLeftOperand() and
+      result = getRightOperand().getFirstInstruction()
+    ) or
+    (
+      child = getRightOperand() and
+      result = getConditionContext().getChildTrueSuccessor(this)
+    )
+  }
+
+  override Instruction getChildFalseSuccessor(TranslatedCondition child) {
+    (child = getLeftOperand() or child = getRightOperand()) and
+    result = getConditionContext().getChildFalseSuccessor(this)
+  }
+}
+
+class TranslatedLogicalOrExpr extends TranslatedBinaryLogicalOperation {
+  override LogicalOrExpr expr;
+
+  override Instruction getChildTrueSuccessor(TranslatedCondition child) {
+    (child = getLeftOperand() or child = getRightOperand()) and
+    result = getConditionContext().getChildTrueSuccessor(this)
+  }
+
+  override Instruction getChildFalseSuccessor(TranslatedCondition child) {
+    (
+      child = getLeftOperand() and
+      result = getRightOperand().getFirstInstruction()
+    ) or
+    (
+      child = getRightOperand() and
+      result = getConditionContext().getChildFalseSuccessor(this)
+    )
+  }
+}
+
+class TranslatedValueCondition extends TranslatedCondition,
+    TTranslatedValueCondition {
+  TranslatedValueCondition() {
+    this = TTranslatedValueCondition(expr)
+  }
+
+  override TranslatedElement getChild(int id) {
+    id = 0 and result = getValueExpr()
+  }
+
+  override Instruction getFirstInstruction() {
+    result = getValueExpr().getFirstInstruction()
+  }
+
+  override predicate hasInstruction(Opcode opcode, InstructionTag tag,
+      Type resultType, boolean isLValue) {
+    tag = ValueConditionConditionalBranchTag() and
+    opcode instanceof Opcode::ConditionalBranch and
+    resultType instanceof VoidType and
```

Comment:
We should use `Language::UnknownType` instead of `VoidType` here (and elsewhere).

Reply:
Fixed in all the places where I used `ObjectType`.

Reply:
(`Language::VoidType` is for instructions that don't return a result at all, including `ConditionalBranch`. `Language::UnknownType` is for instructions that return a result, but we don't know or can't represent its type)